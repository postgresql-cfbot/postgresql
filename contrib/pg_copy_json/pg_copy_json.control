# pg_copy_json extension
comment = 'COPY TO JSON format'
default_version = '1.0'
module_pathname = '$libdir/pg_copy_json'
relocatable = true
