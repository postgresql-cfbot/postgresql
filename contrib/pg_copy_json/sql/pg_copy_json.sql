--
-- COPY TO JSON
--

CREATE EXTENSION pg_copy_json;

-- test copying in JSON format with various styles
-- of embedded line ending characters

create temp table copytest (
	style	text,
	test 	text,
	filler	int);

insert into copytest values('DOS',E'abc\r\ndef',1);
insert into copytest values('Unix',E'abc\ndef',2);
insert into copytest values('Mac',E'abc\rdef',3);
insert into copytest values(E'esc\\ape',E'a\\r\\\r\\\n\\nb',4);

copy copytest to stdout with (format 'json');

-- pg_copy_json do not support COPY FROM
copy copytest from stdout with (format 'json');

-- test copying in JSON format with various styles
-- of embedded escaped characters

create temp table copyjsontest (
    id bigserial,
    f1 text,
    f2 timestamptz);

insert into copyjsontest
  select g.i,
         CASE WHEN g.i % 2 = 0 THEN
           'line with '' in it: ' || g.i::text
         ELSE
           'line with " in it: ' || g.i::text
         END,
         'Mon Feb 10 17:32:01 1997 PST'
  from generate_series(1,5) as g(i);

insert into copyjsontest (f1) values
(E'aaa\"bbb'::text),
(E'aaa\\bbb'::text),
(E'aaa\/bbb'::text),
(E'aaa\bbbb'::text),
(E'aaa\fbbb'::text),
(E'aaa\nbbb'::text),
(E'aaa\rbbb'::text),
(E'aaa\tbbb'::text);

copy copyjsontest to stdout with (format 'json');

-- test force array

copy copytest to stdout (format 'json', force_array);
copy copytest to stdout (format 'json', force_array true);
copy copytest to stdout (format 'json', force_array false);
