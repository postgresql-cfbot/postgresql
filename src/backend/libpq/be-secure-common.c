/*-------------------------------------------------------------------------
 *
 * be-secure-common.c
 *
 * common implementation-independent SSL support code
 *
 * While be-secure.c contains the interfaces that the rest of the
 * communications code calls, this file contains support routines that are
 * used by the library-specific implementations such as be-secure-openssl.c.
 *
 * Portions Copyright (c) 1996-2024, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * IDENTIFICATION
 *	  src/backend/libpq/be-secure-common.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include <sys/stat.h>
#include <unistd.h>

#include "common/percentrepl.h"
#include "common/string.h"
#include "libpq/hba.h"
#include "libpq/libpq.h"
#include "storage/fd.h"
#include "utils/guc.h"
#include "utils/memutils.h"

static HostsLine *parse_hosts_line(TokenizedAuthLine *tok_line, int elevel);

/*
 * Run ssl_passphrase_command
 *
 * prompt will be substituted for %p.  is_server_start determines the loglevel
 * of error messages.
 *
 * The result will be put in buffer buf, which is of size size.  The return
 * value is the length of the actual result.
 */
int
run_ssl_passphrase_command(const char *prompt, bool is_server_start, char *buf, int size)
{
	int			loglevel = is_server_start ? ERROR : LOG;
	char	   *command;
	FILE	   *fh;
	int			pclose_rc;
	size_t		len = 0;

	Assert(prompt);
	Assert(size > 0);
	buf[0] = '\0';

	command = replace_percent_placeholders(ssl_passphrase_command, "ssl_passphrase_command", "p", prompt);

	fh = OpenPipeStream(command, "r");
	if (fh == NULL)
	{
		ereport(loglevel,
				(errcode_for_file_access(),
				 errmsg("could not execute command \"%s\": %m",
						command)));
		goto error;
	}

	if (!fgets(buf, size, fh))
	{
		if (ferror(fh))
		{
			explicit_bzero(buf, size);
			ereport(loglevel,
					(errcode_for_file_access(),
					 errmsg("could not read from command \"%s\": %m",
							command)));
			goto error;
		}
	}

	pclose_rc = ClosePipeStream(fh);
	if (pclose_rc == -1)
	{
		explicit_bzero(buf, size);
		ereport(loglevel,
				(errcode_for_file_access(),
				 errmsg("could not close pipe to external command: %m")));
		goto error;
	}
	else if (pclose_rc != 0)
	{
		char	   *reason;

		explicit_bzero(buf, size);
		reason = wait_result_to_str(pclose_rc);
		ereport(loglevel,
				(errcode_for_file_access(),
				 errmsg("command \"%s\" failed",
						command),
				 errdetail_internal("%s", reason)));
		pfree(reason);
		goto error;
	}

	/* strip trailing newline and carriage return */
	len = pg_strip_crlf(buf);

error:
	pfree(command);
	return len;
}


/*
 * Check permissions for SSL key files.
 */
bool
check_ssl_key_file_permissions(const char *ssl_key_file, bool isServerStart)
{
	int			loglevel = isServerStart ? FATAL : LOG;
	struct stat buf;

	if (stat(ssl_key_file, &buf) != 0)
	{
		ereport(loglevel,
				(errcode_for_file_access(),
				 errmsg("could not access private key file \"%s\": %m",
						ssl_key_file)));
		return false;
	}

	/* Key file must be a regular file */
	if (!S_ISREG(buf.st_mode))
	{
		ereport(loglevel,
				(errcode(ERRCODE_CONFIG_FILE_ERROR),
				 errmsg("private key file \"%s\" is not a regular file",
						ssl_key_file)));
		return false;
	}

	/*
	 * Refuse to load key files owned by users other than us or root, and
	 * require no public access to the key file.  If the file is owned by us,
	 * require mode 0600 or less.  If owned by root, require 0640 or less to
	 * allow read access through either our gid or a supplementary gid that
	 * allows us to read system-wide certificates.
	 *
	 * Note that roughly similar checks are performed in
	 * src/interfaces/libpq/fe-secure-openssl.c so any changes here may need
	 * to be made there as well.  The environment is different though; this
	 * code can assume that we're not running as root.
	 *
	 * Ideally we would do similar permissions checks on Windows, but it is
	 * not clear how that would work since Unix-style permissions may not be
	 * available.
	 */
#if !defined(WIN32) && !defined(__CYGWIN__)
	if (buf.st_uid != geteuid() && buf.st_uid != 0)
	{
		ereport(loglevel,
				(errcode(ERRCODE_CONFIG_FILE_ERROR),
				 errmsg("private key file \"%s\" must be owned by the database user or root",
						ssl_key_file)));
		return false;
	}

	if ((buf.st_uid == geteuid() && buf.st_mode & (S_IRWXG | S_IRWXO)) ||
		(buf.st_uid == 0 && buf.st_mode & (S_IWGRP | S_IXGRP | S_IRWXO)))
	{
		ereport(loglevel,
				(errcode(ERRCODE_CONFIG_FILE_ERROR),
				 errmsg("private key file \"%s\" has group or world access",
						ssl_key_file),
				 errdetail("File must have permissions u=rw (0600) or less if owned by the database user, or permissions u=rw,g=r (0640) or less if owned by root.")));
		return false;
	}
#endif

	return true;
}

/*
 * parse_hosts_line
 *
 * Parses a loaded line from the pg_hosts.conf configuration and pulls out the
 * hostname, certificate, key and CA parts in order to build an SNI config in
 * the TLS backend. Validation of the parsed values is left for the TLS backend
 * to implement.
 */
static HostsLine *
parse_hosts_line(TokenizedAuthLine *tok_line, int elevel)
{
	HostsLine  *parsedline;
	List	   *tokens;
	ListCell   *field;
	AuthToken  *token;

	parsedline = palloc0(sizeof(HostsLine));
	parsedline->sourcefile = pstrdup(tok_line->file_name);
	parsedline->linenumber = tok_line->line_num;
	parsedline->rawline = pstrdup(tok_line->raw_line);

	/* Hostname */
	field = list_head(tok_line->fields);
	tokens = lfirst(field);
	token = linitial(tokens);
	parsedline->hostname = pstrdup(token->string);

	/* SSL Certificate */
	field = lnext(tok_line->fields, field);
	if (!field)
	{
		ereport(elevel,
				errcode(ERRCODE_CONFIG_FILE_ERROR),
				errmsg("missing entry at end of line"),
				errcontext("line %d of configuration file \"%s\"",
						   tok_line->line_num, tok_line->file_name));
		return NULL;
	}
	tokens = lfirst(field);
	token = linitial(tokens);
	parsedline->ssl_cert = pstrdup(token->string);

	/* SSL key */
	field = lnext(tok_line->fields, field);
	if (!field)
	{
		ereport(elevel,
				errcode(ERRCODE_CONFIG_FILE_ERROR),
				errmsg("missing entry at end of line"),
				errcontext("line %d of configuration file \"%s\"",
						   tok_line->line_num, tok_line->file_name));
		return NULL;
	}
	tokens = lfirst(field);
	token = linitial(tokens);
	parsedline->ssl_key = pstrdup(token->string);

	/* SSL CA */
	field = lnext(tok_line->fields, field);
	if (!field)
	{
		ereport(elevel,
				errcode(ERRCODE_CONFIG_FILE_ERROR),
				errmsg("missing entry at end of line"),
				errcontext("line %d of configuration file \"%s\"",
						   tok_line->line_num, tok_line->file_name));
		return NULL;
	}
	tokens = lfirst(field);
	token = linitial(tokens);
	parsedline->ssl_ca = pstrdup(token->string);

	return parsedline;
}

/*
 * load_hosts
 *
 * Reads pg_hosts.conf and passes back a List of parsed lines, or NIL in case
 * of errors.
 */
List *
load_hosts(void)
{
	FILE	   *file;
	ListCell   *line;
	List	   *hosts_lines = NIL;
	List	   *parsed_lines = NIL;
	HostsLine  *newline;
	bool		ok = true;
	MemoryContext oldcxt;
	MemoryContext hostcxt;

	file = open_auth_file(HostsFileName, LOG, 0, NULL);
	if (file == NULL)
	{
		/* An error has already been logged so no need to add one here. */
		return NIL;
	}

	tokenize_auth_file(HostsFileName, file, &hosts_lines, LOG, 0);

	hostcxt = AllocSetContextCreate(PostmasterContext,
									"hosts file parser context",
									ALLOCSET_SMALL_SIZES);
	oldcxt = MemoryContextSwitchTo(hostcxt);

	foreach(line, hosts_lines)
	{
		TokenizedAuthLine *tok_line = (TokenizedAuthLine *) lfirst(line);

		if (tok_line->err_msg != NULL)
		{
			ok = false;
			continue;
		}

		if ((newline = parse_hosts_line(tok_line, LOG)) == NULL)
		{
			ok = false;
			continue;
		}

		parsed_lines = lappend(parsed_lines, newline);
	}

	free_auth_file(file, 0);
	MemoryContextSwitchTo(oldcxt);

	/*
	 * If we didn't find any SNI configuration then that's not an error since
	 * the pg_hosts file is additive to the default SSL configuration.
	 */
	if (ok && parsed_lines == NIL)
	{
		ereport(DEBUG1,
				errmsg("no SNI configuration added from configuration file  \"%s\"",
					   HostsFileName));
		MemoryContextDelete(hostcxt);
		return NIL;
	}

	if (!ok)
	{
		MemoryContextDelete(hostcxt);
		return NIL;
	}

	return parsed_lines;
}
