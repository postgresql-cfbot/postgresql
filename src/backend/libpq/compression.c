/*-------------------------------------------------------------------------
*
 * compression.c
 *	  Functions and variables to support backend configuration of libpq
 *
 *
 * Copyright (c) 2023, PostgreSQL Global Development Group
 *
 * src/backend/libpq/compression.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"
#include "miscadmin.h"
#include "libpq/libpq-be.h"
#include "libpq/compression.h"
#include "utils/guc_hooks.h"

/* GUC variable containing the allowed compression algorithms list (separated by semicolon) */
char	   *libpq_compress_algorithms = "on";
pg_compress_specification libpq_compressors[COMPRESSION_ALGORITHM_COUNT];
size_t		libpq_n_compressors = 0;

bool
check_libpq_compression(char **newval, void **extra, GucSource source)
{
	pg_compress_specification compressors[COMPRESSION_ALGORITHM_COUNT];
	size_t		n_compressors;
	char	   *serialized_compressors;

	if (zpq_parse_compression_setting(*newval, compressors, &n_compressors) == -1)
	{
		GUC_check_errdetail("Cannot parse the libpq_compression setting.");
		return false;
	}

	if (n_compressors > 0)
	{
		guc_free(*newval);
		serialized_compressors = zpq_serialize_compressors(compressors, n_compressors);
		*newval = guc_strdup(ERROR, serialized_compressors);
		pfree(serialized_compressors);
	}
	else
	{
		guc_free(*newval);
		*newval = guc_strdup(ERROR, "");
	}
	return true;
}

void
assign_libpq_compression(const char *newval, void *extra)
{
	if (strlen(newval) == 0)
	{
		libpq_n_compressors = 0;
		return;
	}
	zpq_parse_compression_setting(newval, libpq_compressors, &libpq_n_compressors);
}

void
configure_libpq_compression(Port *port, const char *newval)
{
	pg_compress_specification fe_compressors[COMPRESSION_ALGORITHM_COUNT];
	size_t		n_fe_compressors;

	Assert(!port->zpq_stream);

	if (libpq_n_compressors == 0)
	{
		return;
	}

	/* Init compression */
	port->zpq_stream = zpq_create(libpq_compressors, libpq_n_compressors, MyProcPort->io_stream);
	if (!port->zpq_stream)
	{
		ereport(FATAL,
				errcode(ERRCODE_INTERNAL_ERROR),
				errmsg("failed to initialize the compression stream"));
	}

	if (strlen(newval) == 0)
	{
		return;
	}

	if (!zpq_deserialize_compressors(newval, fe_compressors, &n_fe_compressors))
	{
		ereport(FATAL,
				errcode(ERRCODE_INTERNAL_ERROR),
				errmsg("Cannot parse the _pq_.libpq_compression setting."));
	}

	if (MyProcPort && MyProcPort->zpq_stream)
	{
		pg_compress_algorithm algorithms[COMPRESSION_ALGORITHM_COUNT];
		size_t		n_algorithms = 0;

		if (n_fe_compressors == 0)
		{
			return;
		}

		/*
		 * Intersect client and server compressors to determine the final list
		 * of the supported compressors. O(N^2) is negligible because of a
		 * small number of the compression methods.
		 */
		for (size_t i = 0; i < libpq_n_compressors; i++)
		{
			for (size_t j = 0; j < n_fe_compressors; j++)
			{
				if (libpq_compressors[i].algorithm == fe_compressors[j].algorithm && libpq_compressors[i].compress && fe_compressors[j].decompress)
				{
					algorithms[n_algorithms] = libpq_compressors[i].algorithm;
					n_algorithms += 1;
					break;
				}
			}
		}

		zpq_enable_compression(MyProcPort->zpq_stream, algorithms, n_algorithms);
	}
}
