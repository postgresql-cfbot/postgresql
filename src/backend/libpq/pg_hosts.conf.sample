# PostgreSQL SNI Hostname mappings
# ================================

# HOSTNAME       SSL CERTIFICATE             SSL KEY

