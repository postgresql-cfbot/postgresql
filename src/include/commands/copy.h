/*-------------------------------------------------------------------------
 *
 * copy.h
 *	  Definitions for using the POSTGRES copy command.
 *
 *
 * Portions Copyright (c) 1996-2024, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/commands/copy.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef COPY_H
#define COPY_H

#include "commands/copyapi.h"
#include "nodes/parsenodes.h"
#include "parser/parse_node.h"
#include "tcop/dest.h"

extern void DoCopy(ParseState *pstate, const CopyStmt *stmt,
				   int stmt_location, int stmt_len,
				   uint64 *processed);

extern void ProcessCopyOptions(ParseState *pstate, CopyFormatOptions *opts_out, bool is_from, List *options);
extern CopyFromState BeginCopyFrom(ParseState *pstate, Relation rel, Node *whereClause,
								   const char *filename,
								   bool is_program, copy_data_source_cb data_source_cb, List *attnamelist, List *options);
extern void EndCopyFrom(CopyFromState cstate);
extern bool NextCopyFrom(CopyFromState cstate, ExprContext *econtext,
						 Datum *values, bool *nulls);
extern void CopyFromErrorCallback(void *arg);
extern char *CopyLimitPrintoutLength(const char *str);

extern uint64 CopyFrom(CopyFromState cstate);

extern DestReceiver *CreateCopyDestReceiver(void);

/*
 * internal prototypes
 */
extern CopyToState BeginCopyTo(ParseState *pstate, Relation rel, RawStmt *raw_query,
							   Oid queryRelId, const char *filename, bool is_program,
							   copy_data_dest_cb data_dest_cb, List *attnamelist, List *options);
extern void EndCopyTo(CopyToState cstate);
extern uint64 DoCopyTo(CopyToState cstate);
extern List *CopyGetAttnums(TupleDesc tupDesc, Relation rel,
							List *attnamelist);

#endif							/* COPY_H */
