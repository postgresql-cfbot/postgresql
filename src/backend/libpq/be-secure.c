/*-------------------------------------------------------------------------
 *
 * be-secure.c
 *	  functions related to setting up a secure connection to the frontend.
 *	  Secure connections are expected to provide confidentiality,
 *	  message integrity and endpoint authentication.
 *
 *
 * Portions Copyright (c) 1996-2023, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 *
 * IDENTIFICATION
 *	  src/backend/libpq/be-secure.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include <signal.h>
#include <fcntl.h>
#include <ctype.h>
#include <sys/socket.h>
#include <netdb.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>

#include "libpq/libpq.h"
#include "miscadmin.h"
#include "pgstat.h"
#include "storage/ipc.h"
#include "storage/proc.h"
#include "tcop/tcopprot.h"
#include "utils/memutils.h"

char	   *ssl_library;
char	   *ssl_cert_file;
char	   *ssl_key_file;
char	   *ssl_ca_file;
char	   *ssl_crl_file;
char	   *ssl_crl_dir;
char	   *ssl_dh_params_file;
char	   *ssl_passphrase_command;
bool		ssl_passphrase_command_supports_reload;

#ifdef USE_SSL
bool		ssl_loaded_verify_locations = false;
#endif

/* GUC variable controlling SSL cipher list */
char	   *SSLCipherSuites = NULL;

/* GUC variable for default ECHD curve. */
char	   *SSLECDHCurve;

/* GUC variable: if false, prefer client ciphers */
bool		SSLPreferServerCiphers;

int			ssl_min_protocol_version = PG_TLS1_2_VERSION;
int			ssl_max_protocol_version = PG_TLS_ANY;

/* ------------------------------------------------------------ */
/*			 Procedures common to all secure sessions			*/
/* ------------------------------------------------------------ */

/*
 *	Initialize global context.
 *
 * If isServerStart is true, report any errors as FATAL (so we don't return).
 * Otherwise, log errors at LOG level and return -1 to indicate trouble,
 * preserving the old SSL state if any.  Returns 0 if OK.
 */
int
secure_initialize(bool isServerStart)
{
#ifdef USE_SSL
	return be_tls_init(isServerStart);
#else
	return 0;
#endif
}

/*
 *	Destroy global context, if any.
 */
void
secure_destroy(void)
{
#ifdef USE_SSL
	be_tls_destroy();
#endif
}

/*
 * Indicate if we have loaded the root CA store to verify certificates
 */
bool
secure_loaded_verify_locations(void)
{
#ifdef USE_SSL
	return ssl_loaded_verify_locations;
#else
	return false;
#endif
}

/*
 *	Attempt to negotiate secure session.
 */
int
secure_open_server(Port *port)
{
	int			r = 0;

#ifdef USE_SSL
	r = be_tls_open_server(port);

	ereport(DEBUG2,
			(errmsg_internal("SSL connection from DN:\"%s\" CN:\"%s\"",
							 port->peer_dn ? port->peer_dn : "(anonymous)",
							 port->peer_cn ? port->peer_cn : "(anonymous)")));
#endif

	return r;
}
