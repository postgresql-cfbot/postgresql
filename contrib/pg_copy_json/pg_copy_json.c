/*--------------------------------------------------------------------------
 *
 * pg_copy_json.c
 *		COPY TO JSON (JavaScript Object Notation) format.
 *
 * Portions Copyright (c) 2024, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		contrib/test_copy_format.c
 *
 * -------------------------------------------------------------------------
 */

#include "postgres.h"

#include "commands/copy.h"
#include "commands/defrem.h"
#include "funcapi.h"
#include "libpq/libpq.h"
#include "libpq/pqformat.h"
#include "utils/json.h"

PG_MODULE_MAGIC;

typedef struct
{
	/*
	 * Force output of square brackets as array decorations at the beginning
	 * and end of output, with commas between the rows.
	 */
	bool	force_array;
	bool	force_array_specified;
	
	/* need delimiter to start next json array element */
	bool	json_row_delim_needed;
} CopyJsonData;

static inline void
InitCopyJsonData(CopyJsonData *p)
{
	Assert(p);
	p->force_array = false;
	p->force_array_specified = false;
	p->json_row_delim_needed = false;
}

static void
CopyToJsonSendEndOfRow(CopyToState cstate)
{
	switch (cstate->copy_dest)
	{
		case COPY_DEST_FILE:
			/* Default line termination depends on platform */
#ifndef WIN32
			CopySendChar(cstate, '\n');
#else
			CopySendString(cstate, "\r\n");
#endif
			break;
		case COPY_DEST_FRONTEND:
			/* The FE/BE protocol uses \n as newline for all platforms */
			CopySendChar(cstate, '\n');
			break;
		default:
			break;
	}
	CopyToStateFlush(cstate);
}

static bool
CopyToJsonProcessOption(CopyToState cstate, DefElem *defel)
{
	CopyJsonData	   *p;

	if (cstate->opaque == NULL)
	{
		MemoryContext oldcontext;
		oldcontext = MemoryContextSwitchTo(cstate->copycontext);
		cstate->opaque = palloc0(sizeof(CopyJsonData));
		MemoryContextSwitchTo(oldcontext);
		InitCopyJsonData(cstate->opaque);
	}

	p = (CopyJsonData *)cstate->opaque;

	if (strcmp(defel->defname, "force_array") == 0)
	{
		if (p->force_array_specified)
			ereport(ERROR,
					errcode(ERRCODE_SYNTAX_ERROR),
					errmsg("CopyToJsonProcessOption: redundant options \"%s\"=\"%s\"",
						   defel->defname, defGetString(defel)));
		p->force_array_specified = true;
		p->force_array = defGetBoolean(defel);

		return true;
	}

	return false;
}

static void
CopyToJsonSendCopyBegin(CopyToState cstate)
{
	StringInfoData buf;
	int16		format = 0;

	pq_beginmessage(&buf, PqMsg_CopyOutResponse);
	pq_sendbyte(&buf, format);	/* overall format */
	/*
	 * JSON mode is always one non-binary column
	 */
	pq_sendint16(&buf, 1);
	pq_sendint16(&buf, 0);
	pq_endmessage(&buf);
}

static void
CopyToJsonStart(CopyToState cstate, TupleDesc tupDesc)
{
	CopyJsonData	   *p;

	if (cstate->opaque == NULL)
	{
		MemoryContext oldcontext;
		oldcontext = MemoryContextSwitchTo(cstate->copycontext);
		cstate->opaque = palloc0(sizeof(CopyJsonData));
		MemoryContextSwitchTo(oldcontext);
		InitCopyJsonData(cstate->opaque);
	}

	/* No need to alloc cstate->out_functions */

	p = (CopyJsonData *)cstate->opaque;

	/* If FORCE_ARRAY has been specified send the open bracket. */
	if (p->force_array)
	{
		CopySendChar(cstate, '[');
		CopyToJsonSendEndOfRow(cstate);
	}
}

static void
CopyToJsonOneRow(CopyToState cstate, TupleTableSlot *slot)
{
	Datum				rowdata;
	StringInfo			result;
	CopyJsonData	   *p;

	Assert(cstate->opaque);
	p = (CopyJsonData *)cstate->opaque;

	if(!cstate->rel)
	{
		for (int i = 0; i < slot->tts_tupleDescriptor->natts; i++)
		{
			/* Flat-copy the attribute array */
			memcpy(TupleDescAttr(slot->tts_tupleDescriptor, i),
			TupleDescAttr(cstate->queryDesc->tupDesc, i),
							1 * sizeof(FormData_pg_attribute));
		}
		BlessTupleDesc(slot->tts_tupleDescriptor);
	}
	rowdata = ExecFetchSlotHeapTupleDatum(slot);
	result = makeStringInfo();
	composite_to_json(rowdata, result, false);

	if (p->json_row_delim_needed)
		CopySendChar(cstate, ',');
	else if (p->force_array)
	{
		/* first row needs no delimiter */
		CopySendChar(cstate, ' ');
		p->json_row_delim_needed = true;
	}
	CopySendData(cstate, result->data, result->len);
	CopyToJsonSendEndOfRow(cstate);
}

static void
CopyToJsonEnd(CopyToState cstate)
{
	CopyJsonData	   *p;

	Assert(cstate->opaque);
	p = (CopyJsonData *)cstate->opaque;

	/* If FORCE_ARRAY has been specified send the close bracket. */
	if (p->force_array)
	{
		CopySendChar(cstate, ']');
		CopyToJsonSendEndOfRow(cstate);
	}
}

static const CopyToRoutine CopyToRoutineJson = {
	.type = T_CopyToRoutine,
	.CopyToProcessOption = CopyToJsonProcessOption,
	.CopyToSendCopyBegin = CopyToJsonSendCopyBegin,
	.CopyToStart = CopyToJsonStart,
	.CopyToOneRow = CopyToJsonOneRow,
	.CopyToEnd = CopyToJsonEnd,
};

PG_FUNCTION_INFO_V1(copy_json);
Datum
copy_json(PG_FUNCTION_ARGS)
{
	bool		is_from = PG_GETARG_BOOL(0);

	if (is_from)
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("cannot use JSON mode in COPY FROM")));

	PG_RETURN_POINTER(&CopyToRoutineJson);
}
