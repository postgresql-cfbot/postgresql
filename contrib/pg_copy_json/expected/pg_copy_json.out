--
-- COPY TO JSON
--
CREATE EXTENSION pg_copy_json;
-- test copying in JSON format with various styles
-- of embedded line ending characters
create temp table copytest (
	style	text,
	test 	text,
	filler	int);
insert into copytest values('DOS',E'abc\r\ndef',1);
insert into copytest values('Unix',E'abc\ndef',2);
insert into copytest values('Mac',E'abc\rdef',3);
insert into copytest values(E'esc\\ape',E'a\\r\\\r\\\n\\nb',4);
copy copytest to stdout with (format 'json');
{"style":"DOS","test":"abc\r\ndef","filler":1}
{"style":"Unix","test":"abc\ndef","filler":2}
{"style":"Mac","test":"abc\rdef","filler":3}
{"style":"esc\\ape","test":"a\\r\\\r\\\n\\nb","filler":4}
-- pg_copy_json do not support COPY FROM
copy copytest from stdout with (format 'json');
ERROR:  cannot use JSON mode in COPY FROM
-- test copying in JSON format with various styles
-- of embedded escaped characters
create temp table copyjsontest (
    id bigserial,
    f1 text,
    f2 timestamptz);
insert into copyjsontest
  select g.i,
         CASE WHEN g.i % 2 = 0 THEN
           'line with '' in it: ' || g.i::text
         ELSE
           'line with " in it: ' || g.i::text
         END,
         'Mon Feb 10 17:32:01 1997 PST'
  from generate_series(1,5) as g(i);
insert into copyjsontest (f1) values
(E'aaa\"bbb'::text),
(E'aaa\\bbb'::text),
(E'aaa\/bbb'::text),
(E'aaa\bbbb'::text),
(E'aaa\fbbb'::text),
(E'aaa\nbbb'::text),
(E'aaa\rbbb'::text),
(E'aaa\tbbb'::text);
copy copyjsontest to stdout with (format 'json');
{"id":1,"f1":"line with \" in it: 1","f2":"1997-02-10T17:32:01-08:00"}
{"id":2,"f1":"line with ' in it: 2","f2":"1997-02-10T17:32:01-08:00"}
{"id":3,"f1":"line with \" in it: 3","f2":"1997-02-10T17:32:01-08:00"}
{"id":4,"f1":"line with ' in it: 4","f2":"1997-02-10T17:32:01-08:00"}
{"id":5,"f1":"line with \" in it: 5","f2":"1997-02-10T17:32:01-08:00"}
{"id":1,"f1":"aaa\"bbb","f2":null}
{"id":2,"f1":"aaa\\bbb","f2":null}
{"id":3,"f1":"aaa/bbb","f2":null}
{"id":4,"f1":"aaa\bbbb","f2":null}
{"id":5,"f1":"aaa\fbbb","f2":null}
{"id":6,"f1":"aaa\nbbb","f2":null}
{"id":7,"f1":"aaa\rbbb","f2":null}
{"id":8,"f1":"aaa\tbbb","f2":null}
-- test force array
copy copytest to stdout (format 'json', force_array);
[
 {"style":"DOS","test":"abc\r\ndef","filler":1}
,{"style":"Unix","test":"abc\ndef","filler":2}
,{"style":"Mac","test":"abc\rdef","filler":3}
,{"style":"esc\\ape","test":"a\\r\\\r\\\n\\nb","filler":4}
]
copy copytest to stdout (format 'json', force_array true);
[
 {"style":"DOS","test":"abc\r\ndef","filler":1}
,{"style":"Unix","test":"abc\ndef","filler":2}
,{"style":"Mac","test":"abc\rdef","filler":3}
,{"style":"esc\\ape","test":"a\\r\\\r\\\n\\nb","filler":4}
]
copy copytest to stdout (format 'json', force_array false);
{"style":"DOS","test":"abc\r\ndef","filler":1}
{"style":"Unix","test":"abc\ndef","filler":2}
{"style":"Mac","test":"abc\rdef","filler":3}
{"style":"esc\\ape","test":"a\\r\\\r\\\n\\nb","filler":4}
